cmd/object-handlers.go

```diff
@@ -770,6 +770,9 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 	// encrypted file into an unencrypted one.
 	size := srcInfo.Size
 
+	if globalAutoEncryption && !(crypto.SSEC.IsRequested(r.Header) && crypto.SSECopy.IsRequested(r.Header)) {
+		r.Header.Add(crypto.SSEHeader, crypto.SSEAlgorithmAES256)
```

Comment:
@aead, it is enough to check if destination side has no SSE-C header. Whether source side is encrypted/not is irrelevant when globalAutoEncryption is set.
```
 	if globalAutoEncryption && !(crypto.SSEC.IsRequested(r.Header)  { 
```

Reply:
Oh yeah - I see what you mean. Instead of adding the SSE-S3 header only to valid request add it always if SSE-C headers are missing and even if the request may fail. Yeah, that we can actually do. Got it, changing it :+1:

Reply:
Okay, got confused. Now have the right example ^^
Assume an object `object` encrypted using SSE-C. Client sends COPY request and has set SSE-C headers (`X-Amz-Server-Side...`, ...). The check `globalAutoEncryption && !crypto.SSEC.IsRequested(r.Header)` will return `false` because SSE-C headers are set. However the SSE-C headers refer to the src object not the destination object. The behavior is:
- If the object is encrypted apply SSE-C headers to the source - except SSE-C copy headers are provided (too).
- If the object is not encrypted apply the SSE-C headers to the destination.

So the check `globalAutoEncryption && !crypto.SSEC.IsRequested(r.Header)` will miss the case when the source object is encrypted, SSE-C headers are provided to decrypt it and the destination doesn't get encrypted.

Reply:
Okay, tested this -> AWS returns an error saying: missing SSE-C headers. So you're actually right that if SSE-C headers are present and source is encrypted S3 fails. So the check `globalAutoEncryption && !crypto.SSEC.IsRequested(r.Header)` is sufficient. Chaining this as requested :+1:

Reply:
@aead, just realized - we also need to check if objectAPI.IsEncryptionSupported() here

Reply:
Why? `globalAutoEncryption` will always be false if the backend does not support encryption. See server startup code. Otherwise you would be able to enable auto-encryption but the backend will not support encryption and will ignore set headers...